promql/functions.go

```diff
@@ -286,46 +286,108 @@ func funcIncrease(vals []parser.Value, args parser.Expressions, enh *EvalNodeHel
+	var annos annotations.Annotations
+
+	// Add last 2 float samples if exists
+	for i := len(samples.Floats) - 1; i >= max(0, len(samples.Floats)-2); i-- {
+		LastFourSamples = append(LastFourSamples, Sample{
+			F: samples.Floats[i].F,
+			T: samples.Floats[i].T,
+		})
+	}
+	// Add last 2 histogram samples if exists
+	for i := len(samples.Histograms) - 1; i >= max(0, len(samples.Histograms)-2); i-- {
+		LastFourSamples = append(LastFourSamples, Sample{
+			H: samples.Histograms[i].H,
+			T: samples.Histograms[i].T,
+		})
+	}
+
 	// No sense in trying to compute a rate without at least two points. Drop
 	// this Vector element.
 	// TODO: add RangeTooShortWarning
-	if len(samples.Floats) < 2 {
+	if len(LastFourSamples) < 2 {
 		return out, nil
 	}
+	// sort the last four samples according to their timestamps
+	sort.Slice(LastFourSamples, func(i, j int) bool {
+		return LastFourSamples[i].T < LastFourSamples[j].T
+	})
 
-	lastSample := samples.Floats[len(samples.Floats)-1]
-	previousSample := samples.Floats[len(samples.Floats)-2]
-
-	var resultValue float64
-	if isRate && lastSample.F < previousSample.F {
-		// Counter reset.
-		resultValue = lastSample.F
-	} else {
-		resultValue = lastSample.F - previousSample.F
-	}
-
-	sampledInterval := lastSample.T - previousSample.T
+	lastSample := LastFourSamples[len(LastFourSamples)-1]
+	prevSample := LastFourSamples[len(LastFourSamples)-2]
+	resultSample := lastSample
+	sampledInterval := lastSample.T - prevSample.T
 	if sampledInterval == 0 {
 		// Avoid dividing by 0.
 		return out, nil
 	}
+	switch {
+	case lastSample.H == nil && prevSample.H == nil:
+		if isRate && lastSample.F < prevSample.F {
+			// Counter reset.
+		} else {
+			resultSample.F = lastSample.F - prevSample.F
+		}
+	case lastSample.H != nil && prevSample.H != nil:
+		resultSample.H = lastSample.H.Copy()
+		// irate should only be applied to counters.
+		if isRate && (lastSample.H.CounterResetHint == histogram.GaugeType || prevSample.H.CounterResetHint == histogram.GaugeType) {
+			annos.Add(annotations.NewNativeHistogramNotCounterWarning(metricName, args.PositionRange()))
+		}
+		// idelta should only be applied to gauges.
+		if !isRate && (lastSample.H.CounterResetHint != histogram.GaugeType || prevSample.H.CounterResetHint != histogram.GaugeType) {
+			annos.Add(annotations.NewNativeHistogramNotGaugeWarning(metricName, args.PositionRange()))
+		}
+		_, err := resultSample.H.Sub(prevSample.H)
+		if err != nil {
+			if errors.Is(err, histogram.ErrHistogramsIncompatibleSchema) {
+				return out, annos.Add(annotations.NewMixedExponentialCustomHistogramsWarning(metricName, args.PositionRange()))
+			} else if errors.Is(err, histogram.ErrHistogramsIncompatibleBounds) {
+				return out, annos.Add(annotations.NewIncompatibleCustomBucketsHistogramsWarning(metricName, args.PositionRange()))
+			}
+		}
+		// For counter resets
+		if isRate && lastSample.H.DetectReset(prevSample.H) {
+			_, err := resultSample.H.Add(prevSample.H)
```

Comment:
Instead of subtracting and then adding, let's to the counter reset detection first.

This will allow a result for `irate` even with incompatible buckets because it is simply seen as a counter reset, and the 2nd value is used for the rate. I think that's fine and maybe even to be preferred. (Maybe we should mention that in the docs and the spec as it might be surprising for the user.)

Reply:
Yes, `irate` should have the 2nd value even with the incompatible histograms. But in the current implementation of `rate` also there is no result in case of incompatible histograms in the front and end of the range vector. Both should be consistent right?

Reply:
Another idea is that-
Currently we are returning with no result when an error is encountered in subtraction, but we can do that only in case of it is `idelta` and for `irate` the result will be the 2nd histogram. I've implemented this in my recent push. Please have a look.